tests/test_executor/test_executors.py

```diff
@@ -13,63 +13,93 @@
-        with pytest.raises(EGCGError) as err:
-            e = Executor('dodgy_cmd')
-            e.join()
-            assert 'Command failed: \'dodgy_cmd\'' in str(err)
+        assert self.execute('dodgy_cmd') == 127
+
+    def test_script(self):
+        assert self.execute(os.path.join(self.assets_path, 'countdown.sh')) == 0
+
+    def test_dodgy_script(self):
+        # same exit status as the running script
+        assert self.execute('%s dodgy' % os.path.join(self.assets_path, 'countdown.sh')) == 13
+
+    def test_internal_error(self):
+        with patch.object(self.executor_cls, 'info', side_effect=ValueError('Something went wrong')):
+            with pytest.raises(EGCGError) as err:
+                self.execute('ls')
+
+            assert 'Command failed: ls' in str(err)
 
     def test_process(self):
-        e = Executor('ls ' + os.path.join(self.assets_path, '..'))
+        e = self.executor_cls('ls ' + os.path.dirname(self.assets_path))
         assert e.proc is None
         proc = e._process()
         assert proc is e.proc and isinstance(e.proc, subprocess.Popen)
 
+    def test_bash_syntax(self):
+        with patch.object(self.executor_cls, 'info') as mocked_info:
+            assert self.execute('ls -lh %s | grep __init__' % os.path.dirname(self.assets_path)) == 0
 
-class TestStreamExecutor(TestExecutor):
-    def test_cmd(self):
-        e = StreamExecutor(os.path.join(self.assets_path, 'countdown.sh'))
-        e.start()
-        assert e.join() == 0
+        assert '__init__.py' in mocked_info.call_args_list[1][0][0]
 
-    def test_dodgy_command(self):
-        e = StreamExecutor(os.path.join(self.assets_path, 'countdown.sh') + ' dodgy')
-        e.start()
-        assert e.join() == 13  # same exit status as the running script
 
-    def test_dodgy_cmd(self):
-        with pytest.raises(EGCGError) as err:
-            e = StreamExecutor('dodgy_cmd')
-            e.start()
-            e.error = Mock()
-            e.join()
-            assert 'self.proc command failed: \'dodgy_cmd\'' in str(err)
+class TestStreamExecutor(TestExecutor):
+    executor_cls = StreamExecutor
 
 
 class TestArrayExecutor(TestExecutor):
+    executor_cls = ArrayExecutor
+
+    def execute(self, *args, **kwargs):
+        return super().execute(args, stream=True, **kwargs)
+
     def test_cmd(self):
-        e = ArrayExecutor(['ls', 'ls -lh', 'pwd'], stream=True)
+        e = self.executor_cls(['ls', 'ls -lh', 'pwd'], stream=True)
         e.start()
         assert e.join() == 0
         assert e.exit_statuses == [0, 0, 0]
 
     def test_dodgy_cmd(self):
-        e = ArrayExecutor(['ls', 'non_existent_cmd', 'pwd'], stream=True)
+        e = self.executor_cls(['ls', 'non_existent_cmd', 'pwd'], stream=True)
         for s in e.executors:
             s.error = Mock()
 
         e.error = Mock()
         e.start()
-        with pytest.raises(EGCGError) as err:
-            e.join()
+        assert e.join() == 127
```

Comment:
@mwhamgenomics Is dodgy a library? Again would be best to write a comment explaining why 127 would be satisfactory.

Reply:
The (deliberately) dodgy bit here is `non_existent_cmd` - everything else is valid Bash. The ArrayExecutor should run the 3 commands above, get a list of `[0, 127, 0]` as exit statuses, and return the total, i.e. 127. Will leave a code comment